index.bs

```diff
@@ -52,10 +66,169 @@ dictionary StorageBucketOptions {
+
+1. Set |bucket|'s [=bucket expiration|expiration=] to |expires|.
+
+1. Let |storageBucket| be a new {{StorageBucket}}.
+
+1. Set |storageBucket|'s [=/storage bucket=] to |bucket|.
+
+1. Return |storageBucket|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>validate a bucket name</dfn> given string |name|, run the following steps:
+
+1. If |name| contains any character that is not [=ASCII lower alpha=], [=ASCII digit=], U+005F (_), or U+002D(-), then return failure.
+
+1. If |name| [=string/length=] is 0 or exceeds 64, then return failure.
+
+1. If |name| begins with U+005F (_) or U+002D(-), then return failure.
+
+1. Return.
+
+</div>
+
+When not null, {{StorageBucketOptions/durability}} is a hint to the user agent
+specifying the desired {{StorageBucket/durability}}. The user agent MAY
+create a new {{StorageBucket}} with this [=bucket durability=]. The user agent
+MUST NOT modify the [=bucket durability=] of an existing bucket.
+
 <h3 id="storage-bucket-delete">Deleting a bucket</h3>
+
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">delete(|name|)</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
+
+1. If |storageKey| is failure, then [=exception/throw=] a "{{SecurityError}}" {{DOMException}} and abort these steps.
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. If the result of [=validate a bucket name=] with |name| is failure, then [=/reject=] |p| with an {{InvalidCharacterError}} and abort these steps.
+
+    1. Let |bucket| be the [=/storage bucket=] named |name| in |storageKey| if one exists. Otherwise return.
+
+    1. Remove |bucket|.
+
+    1. [=/Resolve=] |p|.
+
+1. Return |p|.
+
+</div>
+
+Issue: [[Storage]] needs to define bucket removal. The bucket needs to be internally marked as
+removed and subsequent attempts to access it (through associated storage endpoints) should fail.
+
 <h3 id="storage-bucket-keys">Enumerating buckets</h3>
 
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">keys()</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
+
+1. If |storageKey| is failure, then [=exception/throw=] a "{{SecurityError}}" {{DOMException}} and abort these steps.
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. Let |shelf| be the result of running [=obtain a local storage shelf=].
+
+    1. If |shelf| is failure, [=reject=] p with an {{UnknownError}} and abort these steps.
+
+    1. Let |keys| be a new [=/list=].
+
+    1. For each |key| in |shelf|'s [=bucket map=], run the following steps:
+
+        1. Let |bucket| be |shelf|'s [=bucket map=][|key|].
+
+        1. If |bucket|'s [=bucket expiration=] is less than or equal to now, remove |bucket|.
```

Comment:
This feels like a weird spot to be doing expiration.  It feels like the browser should be doing it elsewhere?

Reply:
In our implementation we have a scheduled task for this separately, but we also check upon retrieval to make sure its not expired as well if it hadn't been removed by the tasks yet. Do you think this is an implementation detail that doesn't need to be in the algorithm?